lib/msal-common/src/client/DeviceCodeClient.ts

```diff
@@ -0,0 +1,190 @@
+     * polls token endpoint to exchange device code for tokens
+     * @param request
+     */
+    public async acquireToken(request: DeviceCodeRequest): Promise<string> {
+        this.authority = await this.createAuthority(request.authority);
+        const deviceCodeResponse: DeviceCodeResponse = await this.getDeviceCode(request);
+        request.deviceCodeCallback(deviceCodeResponse);
+        return this.acquireTokenWithDeviceCode(request, deviceCodeResponse);
+    }
+
+    /**
+     * Creates device code request and executes http GET
+     * @param request
+     */
+    private async getDeviceCode(request: DeviceCodeRequest): Promise<DeviceCodeResponse> {
+
+        const deviceCodeUrl = this.createDeviceCodeUrl(request);
+        const headers: Map<string, string> = new Map<string, string>();
+        RequestUtils.addLibrarydataHeaders(headers);
+
+        return this.executeGetRequest(deviceCodeUrl, headers);
+    }
+
+    /**
+     * Executes GET request to device code endpoint
+     * @param deviceCodeUrl
+     * @param headers
+     */
+    private async executeGetRequest(deviceCodeUrl: string, headers: Map<string, string>): Promise<DeviceCodeResponse>{
+
+        let deviceCodeResponse;
+        try {
+            deviceCodeResponse = this.networkClient.sendGetRequestAsync<DeviceCodeResponse>(
+                deviceCodeUrl,
+                {
+                    headers: headers
+                }
+            );
+            return deviceCodeResponse;
+        } catch (error) {
+            console.log(error.response.data); // TODO use logger
+            return error.response.data;
+        }
+    }
+
+    /**
+     * Create device code endpoint url
+     * @param request
+     */
+    private createDeviceCodeUrl(request: DeviceCodeRequest) : string {
+        const params: Map<string, string> = this.createQueryParameters(request);
+        const queryString: string = RequestUtils.createQueryString(params);
+
+        // TODO add device code endpoint to authority class
+        return this.authority.canonicalAuthority + "oauth2/v2.0/devicecode"  + "?" + queryString;
+    }
+
+    /**
+     * Create device code endpoint query parameters
+     * @param request
+     */
+    private createQueryParameters(request: DeviceCodeRequest): Map<string, string>{
+
+        const params: Map<string, string> = new Map<string, string>();
+        RequestUtils.addClientId(params, this.config.authOptions.clientId);
+
+        const scopes = RequestValidator.validateAndGenerateScopes(
+            request.scopes,
+            this.config.authOptions.clientId
+        );
+        RequestUtils.addScopes(params, scopes);
+
+        return params;
+    }
+
+    /**
+     * Creates token request with device code response and polls token endpoint
+     * @param request
+     * @param deviceCodeResponse
+     */
+    private async acquireTokenWithDeviceCode(
```

Comment:
I wonder if this function could be named to denote it is starting a poll for device code